### ISSUE TITLE
symtab/typeArray error on Summit

### ISSUE THREAD

**Issue by hainest** [dyninst/dyninst#1144]
> $ ./test_driver -log tmp.log -v -debugPrint -none -g++ -64 -attach -dynamiclink -pic -test snip_ref_shlib_var

fails with 

>test_driver: /autofs/nccs-svm1_home1/thaines/workspace/summit/paradyn/Xifg8obW/dyninst/src/dyninstAPI/src/BPatch_snippet.C:201: AstNodePtr generateArrayRef(const BPatch_snippet&, const BPatch_snippet&): Assertion `0' failed.

The root cause is that the dynamic cast in `getArrayType` invoked by [BPatch_snippet.C:186](https://github.com/dyninst/dyninst/blob/master/dyninstAPI/src/BPatch_snippet.C#L186) fails when reading the mutatee variable [`srsv3`](https://github.com/dyninst/testsuite/blob/master/src/dyninst/snip_ref_shlib_var_mutatee.c#L59). This would suggest that there is a DWARF problem. However, the DWARF is correct

    <1><c15f>: Abbrev Number: 2 (DW_TAG_base_type)
      24508     <c160>   DW_AT_byte_size   : 1
      24509     <c161>   DW_AT_encoding    : 8  (unsigned char)
      24510     <c162>   DW_AT_name        : (indirect string, offset: 0x5a4): char

    <1><c793>: Abbrev Number: 8 (DW_TAG_array_type)
      25296     <c794>   DW_AT_type        : <0xc15f>
      25297     <c798>   DW_AT_sibling     : <0xc7a2>

    <1><c7a2>: Abbrev Number: 3 (DW_TAG_variable)
      25303     <c7a3>   DW_AT_name        : (indirect string, offset: 0x399e): gv_srsv3
      25304     <c7a7>   DW_AT_decl_file   : 1
      25305     <c7a7>   DW_AT_decl_line   : 59
      25306     <c7a8>   DW_AT_decl_column : 6
      25307     <c7a9>   DW_AT_type        : <0xc793>
      25308     <c7ad>   DW_AT_external    : 1
      25309     <c7ad>   DW_AT_location    : 9 byte block: 3 78 9 3 10 0 0 0 0  (DW_OP_addr: 10030978)

which suggests something is wrong in Symtab. This test did not fail on this platform for the 11.0.1 release. Re-testing that tag *does* fail now. However, there was an OS upgrade on Summit since the 11.0.1 release, so it's hard to compare the two.

I've not seen this failing on other platforms- including other PowerPC systems.

@mxz297 If you have time, could you take a look? It will be next week before I get back to this.

**Comment by mxz297** [comment]
@hainest I just compiled Dyninst and testsuite with their git heads on summit. I cannot reproduce this failure. Below are the modules that I load:

```
module list

Currently Loaded Modules:
  1) lsf-tools/2.0                4) xalt/1.2.1   7) spectrum-mpi/10.4.0.3-20210112    10) intel-tbb-2020.3-gcc-9.1.0-77bppcb
  2) hsi/5.0.2.p5                 5) DefApps      8) cmake/3.21.3                      11) boost-1.77.0-gcc-9.1.0-6amnxfd
  3) darshan-runtime/3.3.0-lite   6) gcc/9.1.0    9) elfutils-0.185-gcc-9.1.0-47ocizb  12) libiberty-2.33.1-gcc-9.1.0-p7sgafg
```

There is a caveat: the gcc module on summit does not seem to apply to `cc`. So if you just `module load` a compiler and the run cmake, you may end up with inconsistent compiler versions between Dyninst and its dependencies. When I built dyninst and testsuite, I explicitly set `-DCMAKE_C_COMPILER=gcc` and `-DCMAKE_CXX_COMPILER=g++` to enforce a consistent compiler version.

**Comment by hainest** [comment]
I'm using the system-provided modules and gcc-11.1.0. Let me try again with gcc-9.1.0.

    1) lsf-tools/2.0                4) xalt/1.2.1   7) spectrum-mpi/10.4.0.3-20210112  10) perl/5.30.1           13) boost/1.77.0      16) elfutils/0.185
    2) hsi/5.0.2.p5                 5) DefApps      8) nano/4.9                        11) python/3.8-anaconda3  14) intel-tbb/2020.3
    3) darshan-runtime/3.3.0-lite   6) gcc/11.1.0   9) git/2.31.1                      12) cmake/3.21.3          15) libiberty/2.33.1

**Comment by hainest** [comment]
>When I built dyninst and testsuite, I explicitly set -DCMAKE_C_COMPILER=gcc and -DCMAKE_CXX_COMPILER=g++ to enforce a consistent compiler version.

I do that as well for precisely the same reason.

**Comment by hainest** [comment]
Summit has been [decommissioned](https://docs.olcf.ornl.gov/systems/2024_olcf_system_changes.html), and I don't see this on any other power machine I test on.
